src/agent/testresultreader.ts

```diff
@@ -0,0 +1,404 @@
+        var defer = Q.defer(); 
+      
+        var testRun2 : ifm.TestRun2;
+        var _this = this;
+
+        xmlreader.read(contents, function (err, res) {
+            if(err) {
+                defer.resolve(err);
+            } 
+            else {
+                try {
+                    testRun2 = _this.parseJUnitXml(res, runContext);
+                    defer.resolve(testRun2);
+                }
+                catch(ex) {
+                    defer.reject(ex);
+                }
+            }
+        });
+
+        return defer.promise;
+    }
+
+    private parseJUnitXml(res, runContext) {
+        var testRun2 : ifm.TestRun2;
+
+        var buildId = runContext.buildId;
+        var buildRequestedFor = runContext.requestedFor;
+        var platform = "";
+        var config = "";
+
+        //read test run summary - runname, host, start time, run duration
+        var runName = "JUnit";
+        var hostName = "";
+        var timeStamp = new Date(); 
+        var totalRunDuration = 0;
+        var totalTestCaseDuration = 0;
+
+        var rootNode = res.testsuite.at(0);
+
+        if(rootNode) {
+
+            if(rootNode.attributes().name) {
+                runName = rootNode.attributes().name;
+            }
+
+            if(rootNode.attributes().hostname) {
+                hostName = rootNode.attributes().hostname;
+            }
+
+            //assume runtimes from xl are current local time since timezone information is not in the xml. If xml date > current local date, fall back to local
+            if(rootNode.attributes().timestamp) {
+                var timestampFromXml = new Date(rootNode.attributes().timestamp);
+                if(timestampFromXml < new Date()) {
+                    timeStamp = timestampFromXml;
+                }                    
+            }
+
+            if(rootNode.attributes().time) {
+                totalRunDuration = rootNode.attributes().time; //in seconds
+            }
+
+            //find test case nodes in JUnit result xml
+            var testResults = [];
+
+            for(var i = 0; i < rootNode.testcase.count(); i ++) {
+                var testCaseNode = rootNode.testcase.at(i);
+
+                //testcase name and type
+                var testName = "";
+                if(testCaseNode.attributes().name) {
+                    testName = testCaseNode.attributes().name;                    
+                } 
+
+                var testStorage = "";
+                if(testCaseNode.attributes().classname) {
+                    testStorage = testCaseNode.attributes().classname;
+                }
+
+                //testcase duration
+                var testCaseDuration = 0; //in seconds
+                if(testCaseNode.attributes().time) {
+                    testCaseDuration = testCaseNode.attributes().time;
+                    totalTestCaseDuration = totalTestCaseDuration + testCaseDuration;
```

Comment:
Not a request for change, this is fine.  You could have saved a few keystrokes with += operator though. :)